### PR TITLE
Initial sound expression support.

### PR DESCRIPTION
**Pull request by microbit-matt-hillsdon**
Adapts code from Microsoft MakeCode (MIT licence).

This was in turn adapted from CODAL (also MIT licensed, (c)
Lancaster University).

Credit and links to source in headers.

### CODE REVIEW

File: src/board/audio.ts [hunk 1]
```diff
@@ -1,6 +1,10 @@
+import { SoundEmojiSynthesizer } from "./audio/sound-emoji-synthesizer";
+import { parseSoundEffects, SoundEffect } from "./audio/sound-expressions";
+
 interface AudioUIOptions {
   defaultAudioCallback: () => void;
   speechAudioCallback: () => void;
+  soundExpressionDoneCallback: () => void;
```
Comment (microbit-matt-hillsdon) [inline src/board/audio.ts]: I don't think this was used in the end.

File: src/board/audio.ts [hunk 2]
```diff
@@ -14,13 +18,20 @@ export class AudioUI {
 
   default: BufferedAudio | undefined;
   speech: BufferedAudio | undefined;
+  soundExpression: BufferedAudio | undefined;
+  soundExpressionDoneCallback: undefined | (() => void);
```
Comment (microbit-matt-hillsdon) [inline src/board/audio.ts]: Not used.

File: src/board/audio.ts [hunk 3]
```diff
@@ -136,3 +191,30 @@ class BufferedAudio {
     source.start(startTime);
   }
 }
+
+function replaceBuiltin(expression: string) {
```
Comment (microbit-matt-hillsdon) [inline src/board/audio.ts]: Should probably go in its own file.

File: src/Makefile [hunk 4]
```diff
@@ -44,7 +44,7 @@ COPT += -O3 -DNDEBUG
 endif
 
 JSFLAGS += -s ASYNCIFY
-JSFLAGS += -s EXPORTED_FUNCTIONS="['_mp_js_main','_mp_js_request_stop','_microbit_hal_audio_ready_callback','_microbit_hal_audio_speech_ready_callback','_microbit_hal_gesture_callback', '_microbit_hal_level_detector_callback']"
+JSFLAGS += -s EXPORTED_FUNCTIONS="['_mp_js_main','_mp_js_request_stop','_microbit_hal_audio_ready_callback','_microbit_hal_audio_speech_ready_callback','_microbit_hal_gesture_callback','_microbit_hal_level_detector_callback']"
```
Comment (microbit-matt-hillsdon) [inline src/Makefile:47]: I changed then reverted this so this is just a whitespace fix now.

File: src/board/audio/built-in-sounds.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+/**
```
Comment (microbit-matt-hillsdon) [inline src/board/audio/built-in-sounds.ts:1]: This is a manual conversion from CODAL by me (not sure where this code ended up in MakeCode).

File: src/board/audio/musical-progressions.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
+/**
```
Comment (microbit-matt-hillsdon) [inline src/board/audio/musical-progressions.ts:1]: This is just the MakeCode code tweaked to compile here and use ESM.

File: src/board/audio/sound-emoji-synthesizer.ts [hunk 7]
```diff
@@ -0,0 +1,253 @@
+/**
```
Comment (microbit-matt-hillsdon) [inline src/board/audio/sound-emoji-synthesizer.ts:1]: This is just the MakeCode code tweaked to compile here and use ESM.

Their TypeScript is older so I had fun with null/undefined fields so there's a bit more change than you might expect.

File: src/board/audio/sound-expressions.ts [hunk 8]
```diff
@@ -0,0 +1,293 @@
+/**
```
Comment (microbit-matt-hillsdon) [inline src/board/audio/sound-expressions.ts:1]: This is just the MakeCode code tweaked to compile here and use ESM.

File: src/board/audio/sound-synthesizer-effects.ts [hunk 9]
```diff
@@ -0,0 +1,200 @@
+/**
```
Comment (microbit-matt-hillsdon) [inline src/board/audio/sound-synthesizer-effects.ts:1]: This is just the MakeCode code tweaked to compile here and use ESM.

File: src/board/audio/sound-synthesizer.ts [hunk 10]
```diff
@@ -0,0 +1,74 @@
+/**
```
Comment (microbit-matt-hillsdon) [inline src/board/audio/sound-synthesizer.ts:1]: This is just the MakeCode code tweaked to compile here and use ESM.

File: src/board/audio/sound-emoji-synthesizer.ts [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+  effectPointer: number = -1;
+  effectBuffer: SoundEffect[] = [];
+
+  get effect() {
+    return this.effectBuffer[this.effectPointer];
+  }
+
+  constructor(
+    id: number,
+    private onDone: () => void,
```
Comment (microbit-matt-hillsdon) [inline src/board/audio/sound-emoji-synthesizer.ts:50]: I added this onDone callback to mirror DEVICE_SOUND_EMOJI_SYNTHESIZER_EVT_DONE that happens in hardware.

**Review by microbit-robert — APPROVED** [review]
LGTM. All works as expected.

**Pull request merged**